### PR TITLE
Avalonia.Input: add keyboard key states query properties to IKeyboardDevice

### LINKED ISSUE #2422: Implement an API to check if CapsLock is on

### PR DESCRIPTION
**Pull request by yatli**
Work in progress.

Fixes #2422

=== TODO ===

- [x] Windows
- [ ] X11
- [ ] OSX
- [ ] Tests

### CODE REVIEW

**Comment by yatli** [comment]
I currently have no idea how to implement this on osx. Looks like there's no single capslock state but each keyboard has its own states.

**Comment by yatli** [comment]
See https://github.com/yatli/fvim/pull/142 for usage.

File: src/Avalonia.Input/IKeyboardDevice.cs [hunk 1]
```diff
@@ -64,5 +64,9 @@ public interface IKeyboardDevice : IInputDevice, INotifyPropertyChanged
             IInputElement element, 
             NavigationMethod method,
             KeyModifiers modifiers);
+
+        KeyStates NumLock { get; }
```
Comment (yatli) [inline src/Avalonia.Input/IKeyboardDevice.cs:68]: perhaps `bool`?

**Comment by kekekeks** [comment]
It's better to embed misc key states in the key event itself rather than extending IKeyboardDevice.

Both IKeyboardDevice and IPointerDevice should only be used for querying immutable device properties (such as pointer type). IKeyboardDevice is here for historical reasons and will probably be removed completely before 1.0 release.

**Comment by yatli** [comment]
I worry there will be a use case where the status query is not bound to an input event.
Suppose we make a terminal, and the user press the scroll lock key to freeze the terminal (like ^s in ssh)

**Comment by kekekeks** [comment]
>Suppose we make a terminal, and the user press the scroll lock key to freeze the terminal

Then we should report the scroll lock state change.

**Comment by kekekeks** [comment]
I guess there are cases when it's the current state of the keyboard keys that matters. However, in general those should be attached to a particular input event, since input events might get processed after keyboard state got changed, i. e. user presses PgDown then toggles ScrollLock. The app might have been busy with processing something on the UI thread, so it would handle PgDown with a wrong key state.

**Comment by kekekeks** [comment]
Another problem is that I don't see any way for querying the current scroll lock status through X11. There is a way to query the current LED status, but there are systems when LED status doesn't match with the actual mode (e. g. I'm using my Caps Lock LED for indicating the current keyboard layout).

**Comment by yatli** [comment]
Yeah, I was thinking of exactly that.
Like, when a lot of data floods into the terminal and the ui loop is busy, the user presses scroll lock to try to chill out. If we wait for the event system to get back, it won't be until the current messages are being processed so the feedback loop will never be closed. Instead, we should be able to check the state from the io thread to decide whether to post more work to the ui thread.

**Comment by yatli** [comment]
I once used a program to turn the leds into hardware monitors (hdd activity etc.) -- in that case the LED state has nothing to do with the lock condition

**Comment by yatli** [comment]
I've done my round of search. I too cannot find information about accessing the capslock state. The Xkbd extension seems to equalize the state and the led status, but then the leds can be controlled with `/sys/something/led/brightness` which would make it inconsistent.

**Comment by kekekeks** [comment]
```
KeyPress event, serial 40, synthetic NO, window 0x13e00001,
    root 0x195, subw 0x0, time 331589103, (120,-19), root:(1032,402),
    state 0x12, keycode 45 (keysym 0x4b, K), same_screen YES,
    XLookupString gives 1 bytes: (4b) "K"
    XmbLookupString gives 1 bytes: (4b) "K"
    XFilterEvent returns: False
```
That's with CapsLock LED off and CapsLock + NumLock behavior being enabled.

NumLock seems to be passed as 0x10 flag in `state`, CapsLock` with 0x2 flag.
Toggling ScrollLock doesn't seem to have any effect on the state field.

**Comment by mcmikecreations** [comment]
I'm looking forward to getting this resolved!

**Comment by jmacato** [comment]
Closing this PR for inactivity. Ping me if this is updated accordingly.

**Pull request closed without merging**